Automattic/Co-Authors-Plus
# Fix/missing wp user type

## Description

I came across this bug when attempting to do the following:
```
$author_data = [
    'wp_user_nicename', // WP User who also happens to be a Guest Author
    'guest_author_only_nicename'
];
$coauthors_plus->add_coauthors( $post_id, $author_data, $append = false );
```

I placed my `WP_User` first in the array because of [this block of code](https://github.com/Automattic/Co-Authors-Plus/blob/424b337dd21a15bd69a4043af1116a6d2811263f/co-authors-plus.php#L1003-L1008). It will update the `wp_posts.post_author` column for the first WP_User it finds.

However, I noticed that the `wp_posts.post_author` column was not being updated as I was expecting.

I started digging through the code and found that this block of code was essentially being skipped, due to the fact that [this function always returns a Guest Author object](https://github.com/Automattic/Co-Authors-Plus/blob/424b337dd21a15bd69a4043af1116a6d2811263f/co-authors-plus.php#L339-L344).

This is actually causing the `add_coauthors` function to always respond with `false` for instances where `wp_posts.post_author` = 0 or where the user no longer exists on the system. Even if you're passing in a valid WP_User who is also a Guest Author, it would respond with `false`, as was my case.

With this fix, we retain the fact that the original `user_nicename` we passed is also a WP_User. In this way, we don't waste compute cycles in that foreach loop where it would never find a valid WP_User it's trying to look for.

## Deploy Notes

Are there any new dependencies added that should be taken into account when deploying to WordPress.org?
No

## Steps to Test

Outline the steps to test and verify the PR here.

You will need 5 authors:

1. WP_User who is not linked to Guest Author ( user_nicename: wp_user_1 )
2. WP_User who is not linked to Guest Author ( user_nicename: wp_user_2 )
3. Guest Author who is not linked ( user_nicename: ga_1 )
4. Guest Author who is not linked ( user_nicename: ga_2 )
5. WP_User who is linked to Guest Author ( user_nicename: wp_user_and_ga_1 )

Example:

1. Check out PR.
2. Instantiate CoAuthor_Plus: `$cap = new CoAuthor_plus();`
3. Execute `$cap->add_coauthors( $post_id, [ 'wp_user_1' ] )`; result should be `true`. `wp_posts.post_author` should be equal to `wp_users.ID`
4. Execute `$cap->add_coauthors( $post_id, [ 'ga_1', 'ga_2' ] )`; result should be `false`. This is because `$append = false` and there was no `wp_user` passed as a CoAuthor. Notice also that `wp_posts.post_author` remains untouched.
6. Execute Step 1 again, then execute Step 4 again, this time passing `$append = true`, like so: `$cap->add_coauthors( $post_id, [ 'ga_1', 'ga_2' ], true )`; result should be `true`. Notice, that all 3 authors are correctly displayed.
7. Execute `$cap->add_coauthors( $post_id, [ 'ga_1', 'wp_user_2' ], true )`; result should be true. Notice that `wp_posts.post_author` _is not_ overwritten. This is because we are appending authors to the post.
8. Execute `$cap->add_coauthors( $post_id, [ 'ga_2', 'wp_user_and_ga_1' ] )`; result should be `true`. Notice here that the WP_User is placed second in the array, but it is correctly picked up now, and the `wp_posts.post_author` column is correctly updated to match the user's `wp_users.ID` column.
